### PR TITLE
Option to update trackers when chapter marked as read

### LINKED ISSUE #619: Add an option to update tracking when marked as read.
### Describe your suggested feature

The app only syncs when I get to the last page of the chapter. That's a problem because when I read online, I can't simply mark the chapter as read to have it update the tracker.

### Other details

Example:
Screenshot_20240404_190305_Ykai

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have written a short but informative title.
- [X] I have updated the app to version **[0.16.4](https://github.com/mihonapp/mihon/releases/latest)**.
- [X] I will fill out all of the requested information in this form.

### PR DESCRIPTION
**Pull request by Animeboynz**
- [ ] Prompt to revert on unread?

Shows a snack bar to update trackers when you mark chapters as read.
Closes #619 

More Context : https://git.mihon.tech/tachiyomi/tachiyomi/issues/389#issuecomment-28053

image

### CODE REVIEW

File: app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt [hunk 1]
```diff
@@ -718,6 +721,27 @@ class MangaScreenModel(
...
+
+                val formattedChapterNumber = if (maxChapterNumber % 1.0 == 0.0) {
+                    maxChapterNumber.toInt().toString()
+                } else {
+                    maxChapterNumber.toString()
+                }
+
+                val result = snackbarHostState.showSnackbar(
+                    message = "Update trackers to chapter $formattedChapterNumber?",
+                    actionLabel = "Yes",
```
Comment (AntsyLich) [inline app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt:736]: ...

Comment (Animeboynz) [reply app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt:736]: Why bother with i18n if people hate the concept, just leaving this PR here to get opinions

Comment (Animeboynz) [reply app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt:736]: Unless you want me to switch it now

Comment (AntsyLich) [reply app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt:736]: Yes

Comment (Animeboynz) [reply app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt]: done now.

Comment (Animeboynz) [reply app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt]: Should I also prompt if the user marks a chapter as unread?

File: app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt [hunk 2]
```diff
@@ -718,6 +721,23 @@ class MangaScreenModel(
...
+
+            if (read) {
+                val maxChapterNumber = chapters.maxOf { it.chapterNumber }
+
+                val formattedChapterNumber = maxChapterNumber.toInt()
+
+                val result = snackbarHostState.showSnackbar(
+                    message = context.stringResource(MR.strings.confirm_tracker_update, formattedChapterNumber),
+                    actionLabel = context.stringResource(MR.strings.action_ok),
+                    duration = SnackbarDuration.Short,
```
Comment (AntsyLich) [inline app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt:733]: Should probably show until dismissed?

Comment (Animeboynz) [reply app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt:733]: In my opinion no, I tried it and it just felt really annoying, that's why I added it. I feel like there are people who won't like this setting and we don't give them an option to disable it so the least we can do is put it on a timer. And the short duration is actually decently long.

Comment (BrutuZ) [reply app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt:733]: > I feel like there are people who won't like this setting and we don't give them an option to disable it so the least we can do is put it on a timer.

Voicing the presence of such users here too o/
No need to be conservative with the "feel like" either, you asked if that would be the case and we answered xD

File: app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt [hunk 3]
```diff
@@ -718,6 +721,23 @@ class MangaScreenModel(
                 read = read,
                 chapters = chapters.toTypedArray(),
             )
+
+            if (read) {
```
Comment (AntsyLich) [inline app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt:725]: Should only show the snackbar if max chapter number is less than what is in track

File: app/src/main/java/eu/kanade/domain/track/interactor/TrackChapter.kt [hunk 4]
```diff
@@ -56,4 +56,18 @@ class TrackChapter(
...
+            return false
+        }
+
+        tracks.map { track ->
+            if (chapterNumber > track.lastChapterRead) {
+                return true
+            }
+        }
+        return false
+    }
```
Comment (AntsyLich) [inline app/src/main/java/eu/kanade/domain/track/interactor/TrackChapter.kt:72]: Move this inside MangaScreenModel

Comment (Animeboynz) [reply app/src/main/java/eu/kanade/domain/track/interactor/TrackChapter.kt]: 👍

File: app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt [hunk 5]
```diff
@@ -718,6 +721,27 @@ class MangaScreenModel(
...
                 chapters = chapters.toTypedArray(),
             )
+
+            val tracks = getTracks.await(mangaId)
+            val maxChapterNumber = chapters.maxOf { it.chapterNumber }
+            val shouldPromptTrackingUpdate = tracks.any { track ->
+                maxChapterNumber > track.lastChapterRead
+            }
+
+            if (read && shouldPromptTrackingUpdate) {
```
Comment (AntsyLich) [inline app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt:731]:
```suggestion
            if (!shouldPromptTrackingUpdate) return@launchIO
```

File: app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt [hunk 6]
```diff
@@ -718,6 +721,27 @@ class MangaScreenModel(
                 read = read,
                 chapters = chapters.toTypedArray(),
             )
+
```
Comment (AntsyLich) [inline app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt:724]:
```suggestion
            if (!read) return@launchIO
```

**Review by AntsyLich — CHANGES_REQUESTED** [review]

File: app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt [hunk 7]
```diff
@@ -718,6 +721,26 @@ class MangaScreenModel(
...
+            if (!read) return@launchIO
```
Comment (AntsyLich) [inline app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt:725]:
```suggestion
            if (!read) return@launchIO
            
```

File: app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt [hunk 8]
```diff
@@ -718,6 +721,26 @@ class MangaScreenModel(
...
+            val tracks = getTracks.await(mangaId)
+            val maxChapterNumber = chapters.maxOf { it.chapterNumber }
+            val shouldPromptTrackingUpdate = tracks.any { track ->
+                maxChapterNumber > track.lastChapterRead
+            }
+
+            if (!shouldPromptTrackingUpdate) return@launchIO
```
Comment (AntsyLich) [inline app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt:732]:
```suggestion
            val shouldPromptTrackingUpdate = tracks.any { maxChapterNumber > it.lastChapterRead }

            if (!shouldPromptTrackingUpdate) return@launchIO
            
```

File: app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt [hunk 9]
```diff
@@ -718,6 +721,26 @@ class MangaScreenModel(
...
+            val formattedChapterNumber = maxChapterNumber.toInt()
+            val result = snackbarHostState.showSnackbar(
+                message = context.stringResource(MR.strings.confirm_tracker_update, formattedChapterNumber),
```
Comment (AntsyLich) [inline app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt:735]:
```suggestion
            val result = snackbarHostState.showSnackbar(
                message = context.stringResource(MR.strings.confirm_tracker_update, maxChapterNumber.toInt()),
```

File: app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt [hunk 10]
```diff
@@ -718,6 +721,26 @@ class MangaScreenModel(
...
+                actionLabel = context.stringResource(MR.strings.action_ok),
+                duration = SnackbarDuration.Short,
+                withDismissAction = true,
+            )
+
+            if (result == SnackbarResult.ActionPerformed) {
+                trackChapter.await(context, mangaId, maxChapterNumber)
+            }
         }
         toggleAllSelection(false)
```
Comment (AntsyLich) [inline app/src/main/java/eu/kanade/tachiyomi/ui/manga/MangaScreenModel.kt:745]: move this at the top

**Review by AntsyLich — CHANGES_REQUESTED** [review]

**Review by AntsyLich — APPROVED** [review]

**Pull request merged**